src/runtime/profiler_common.cpp

```diff
@@ -209,15 +204,48 @@ WEAK halide_profiler_pipeline_stats *halide_profiler_get_pipeline_state(const ch
     return nullptr;
 }
 
-// Returns a token identifying this pipeline instance.
-WEAK int halide_profiler_pipeline_start(void *user_context,
+// Populates the instance state struct
+WEAK int halide_profiler_instance_start(void *user_context,
                                         const char *pipeline_name,
                                         int num_funcs,
-                                        const uint64_t *func_names) {
-    halide_profiler_state *s = halide_profiler_get_state();
+                                        const uint64_t *func_names,
+                                        halide_profiler_instance_state *instance) {
+
+    // Tell the instance where we stashed the per-func state - just after the
+    // instance itself.
+
+    // First check that the layout agrees with the amount of stack space
+    // allocated in the pipeline
+    static_assert((sizeof(halide_profiler_func_stats) & 7) == 0);
+    halide_profiler_func_stats *funcs = (halide_profiler_func_stats *)(instance + 1);
 
+    // Zero initialize the instance and func state
+    memset(instance, 0, (uint8_t *)(funcs + num_funcs) - (uint8_t *)instance);
+
+    instance->funcs = funcs;
+
+    halide_profiler_state *s = halide_profiler_get_state();
     LockProfiler lock(s);
 
+    // Push this instance onto the running instances list
+    if (s->instances) {
+        s->instances->prev_next = &(instance->next);
+    }
+    instance->next = s->instances;
+    instance->prev_next = &(s->instances);
+    s->instances = instance;
+
+    // Find or create the pipeline statistics for this pipeline.
+    halide_profiler_pipeline_stats *p =
+        find_or_create_pipeline(pipeline_name, num_funcs, func_names);
+    if (!p) {
+        // Allocating space to track the statistics failed.
+        return halide_error_out_of_memory(user_context);
+    }
+
+    // Tell the instance the pipeline to which it belongs.
+    instance->pipeline_stats = p;
+
```

Comment:
Here might be a good point to just take the `halide_current_time_ns()` to take the start time, to produce accurate wall times, as opposed to summing samples (which I demonstrated to be very noise/inaccurate in #5796).

Reply:
Will fix in next push, though I won't resolve this because the way I did it might need discussion.

Reply:
Pushed. There was a question about whether or not pipeline start and end should count as sampling events, which could be used to ensure the sum of the billed times for each func equalled the wall clock time of the whole pipeline, but I decided that would actually introduce some biases, so there's an awkward adjustment step at pipeline exit to rescale the billed times to add up to the wall clock time.